### PR TITLE
[rllib] Custom supervised loss API

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

It's already technically possible to mix in supervised losses with any RL algorithm, but not easy to figure out how. This PR adds:
- a runnable example of how to do it
- a conversion function from the offline input reader class to a TF input pipeline (torch users can just call reader.next() directly)

## Related issue number

Closes https://github.com/ray-project/ray/issues/3866

### CODE REVIEW

File: doc/source/rllib-models.rst [hunk 1]
```diff
@@ -8,8 +8,11 @@ The following diagram provides a conceptual overview of data flow between differ
 The components highlighted in green can be replaced with custom user-defined implementations, as described in the next sections. The purple components are RLlib internal, which means they can only be modified by changing the algorithm source code.
 
 
+Default Behaviours
```
Comment (ericl) [inline doc/source/rllib-models.rst]: Moved some sections around to clean this page up.

File: python/ray/rllib/models/model.py [hunk 2]
```diff
@@ -165,15 +201,25 @@ def _validate_output_shape(self):
...
-def _restore_original_dimensions(input_dict, obs_space, tensorlib=tf):
+@DeveloperAPI
+def restore_original_dimensions(obs, obs_space, tensorlib=tf):
+    """Unpacks Dict and Tuple space observations into their original form.
+
+    This is needed since we flatten Dict and Tuple observations in transit.
+    Before sending them to the model though, we should unflatten them into
+    Dicts or Tuples of tensors.
+
+    Arguments:
```
Comment (richardliaw) [inline python/ray/rllib/models/model.py]: Returns

File: python/ray/rllib/offline/input_reader.py [hunk 3]
```diff
@@ -17,3 +25,100 @@ def next(self):
...
+        threading.Thread.__init__(self)
+        self.sess = tf.get_default_session()
+        self.daemon = True
+        self.input_reader = input_reader
+        self.keys = keys
+        self.queue = queue
+        self.placeholders = [tf.placeholder(dtype) for dtype in dtypes]
+        self.enqueue_op = queue.enqueue(
+            {k: ph
+             for k, ph in zip(keys, self.placeholders)})
```
Comment (richardliaw) [inline python/ray/rllib/offline/input_reader.py]: nit - dict(zip

File: python/ray/rllib/offline/input_reader.py [hunk 4]
```diff
@@ -17,3 +25,100 @@ def next(self):
...
+
+    def enqueue(self, batch):
+        data = {
+            self.placeholders[i]: batch[self.keys[i]]
+            for i in range(len(self.keys))
```
Comment (richardliaw) [inline python/ray/rllib/offline/input_reader.py]: nit: for i, key in enumerate(self.keys)

File: python/ray/rllib/offline/input_reader.py [hunk 5]
```diff
@@ -17,3 +25,100 @@ def next(self):
             SampleBatch or MultiAgentBatch read.
         """
         raise NotImplementedError
+
+    @PublicAPI
+    def tf_input_ops(self, queue_size=1, dequeue_n=1):
```
Comment (richardliaw) [inline python/ray/rllib/offline/input_reader.py]: is dequeue_n used?

**Review by richardliaw — APPROVED** [review]
left comments, looks good

**Pull request merged**